Allow for extensions on a per filesystem view basis

obreitwi opened the pull request:
Resolves #1789

In our scenario we have a a globally installed (i.e., read-only) set of
packages from which each user should then be able to add a subset to
local filesystem views. However, the current extension mechanism only
allows for one set of extensions to be activated for any given extendee.

One way to circumvent this issue is to use a module system, but this way
environment variables tend to get very long after a short while.
Furthermore, since an error is raised when two files have the same
target in the view, the user can immediately spot when multiple specs
from one package are loaded (an error that sill creeps up from time to
time).

In order to support multiple (and maybe mutually exclusive)
configurations of extensions for a given extendee, the following changes
were made. The aim is to silently activate any specified extensions when
adding them to the filesystem view; no actions from the user required.

TL;DR: Adding python packages built with py-setuptools to fileystem
views without globally activating them should "just work".

* `DirectoryLayout`/`Package`: All extension-related command accept a
`path_view` argument restricting extension operations to the given
filesystem view.

* `Package`: `{do_,}{de,}activate`-functions also accept a `path_view`
argument. If `path_view` is `not None`, the necessary extension files
are linked into the view instead of the extendee's install directory.

* `LinkTree`: `merge`-method now accepts a link argument to allow for
both symlinks and hardlinks.

* `view`: All linking operations have been refactored to use `LinkTree`s
instead of custom code.

* `view`: Check extensions prior to linking packages, activate
extensions in filesystem view afterwards.

* `view statlink`: The `--verbose` option now prints the `short_spec` of
packages linked.

* `view statlink`: `status` and `remove` now accept `--all` argument to
display the status of/remove all packages in the filesystem view.

* `view statlink`: Now uses `display_specs`-function for prettier
output.

* `view remove`: Only removes packages specified by user, packages
depending on those packages and packages that only user-specified
packages depend on.

I am currently working on some meaningful tests (I did not find any for the `view` command on its own to start from), but would appreciate some feedback in the meantime.

paulhopkins commented:
This looks broadly similar to my code mentioned in #3245. One thing I did add later was removing absolute references from merged `easy-install.pth` as dependencies were activated anyway and the file ends up with lots of duplicates. Maybe this does that as well, but I did not see it.

citibeth commented:
This is clearly the right direction.  `spack activate` fundamentally breaks Spack, for the reasons given above; and this is a way to do a similar thing that does not break Spack.  (I personally have been happy with loading modules).

muffgaga commented:
@citibeth wrote:
> [...] (I personally have been happy with loading modules).

We started to get quite unhappy when we crossed the 50 entries per environment variable barrier (we basically dropped all system-installed software and rely 100% on spack for all our software dependencies).
→ You cannot open manpages anymore, loading python gets incredible slow, …

Due to @obreitwi's change we (@electronicvisions) are all happy again ;)

citibeth commented:
Hmm.... I believe I've passed that threshold already, but I'm not unhappy.
Maybe it depends on the fileserver being used, I'm running off a local
drive.

On Tue, Feb 28, 2017 at 10:14 AM, Eric Müller <notifications@github.com>
wrote:

> @citibeth <https://github.com/citibeth> wrote:
>
> [...] (I personally have been happy with loading modules).
>
> We started to get quite unhappy when we crossed the 50 entries per
> environment variable barrier (we basically dropped all system-installed
> software and rely 100% on spack for all our software dependencies).
> → You cannot open manpages anymore, loading python gets incredible slow, …
>
> Due to @obreitwi <https://github.com/obreitwi>'s change we (
> @electronicvisions <https://github.com/electronicvisions>) are all happy
> again ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3227#issuecomment-283066480>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdxzKPPRiZyxcUmrA7tRXIXcBZUmmks5rhDnwgaJpZM4MKE5G>
> .
>

obreitwi commented:
@citibeth yeah, we use NFS and it is unbearable...

Just added a missing feature that locally and globally activated extensions are handled correctly (forgot about this as we do not plan on using any globally activated extensions).

@paulhopkins how about using `os.path.realpath` to normalize all paths, make sure there are no duplicates and finally `os.path.relpath` to generatre a _pretty_ `easy_install.pth` file?

obreitwi commented:
Some further functionality (plus a rebase on current `develop` branch):
* Added `-i/--ignore-conflicts` command line option to print warnings
  instead of raising an exception whenever conflicting files are found.
  Please note that this only works for linking non-extensions as the
  activation-procedure should take care of conflicting files in
  extensions.

* To that end, LinkTree now has the option to `report_conflicts` instead
  of raising assertions when `merge`-ing.

* When activating a python extension, the `easy_install.pth` file is
  pruned from duplicated lines (while preserving the original order).

obreitwi commented:
Actually, I just noticed #3587, which solves the problem of filesystem views being unusable with python extensions in general, still, we need this PR to have different sets of activated extensions in each view within one spack installation.

scheibelp commented:
I've spoken with @tgamblin and I will be reviewing this over the next couple days.

Two initial notes:

* I notice you generally check for things like ```if path_view is None``` vs. ```if not path_view``` - are there times when the empty string would be a valid path_view?
* When I build "git" in your PR and then try ```spack view soft sandbox/ git``` I get the error ```Specs passed to a DirectoryLayout must be concrete!```. I don't think this is on account of the merge conflict (that appears to just be imports). The original view logic appears to be able to do this. If you're not sure I can look into it myself.

obreitwi commented:
Thank you for taking the time.

* You are correct, there is no point in differentiating between `None` and `""`, an empty path is never a valid location. I have refactored accordingly.

* We did, however, encounter something similar a while back, but I did not find the time to properly investigate:
In our case, the error occurred because the `.extendee_spec` of a given package was still present, even though the package was not configured to be an extension. If I understood the code correctly, this is not supposed to happen.
Upon (quick) second investigation it might also be the case that the [`flatten(...)`](https://github.com/electronicvisions/spack/blob/03724ff384f0f03eb55a44437d7176cad3306f95/lib/spack/spack/cmd/view.py#L181)-method should simply use `concretized()` instead…

I will test this out as soon as I have time!

scheibelp commented:
> When I build "git" in your PR and then try ```spack view soft sandbox/ git``` I get the error ```Specs passed to a DirectoryLayout must be concrete!.```

If you modify check_is_extension like so then this issue should be resolved:

```
    if not ext_spec.package.is_extension:
        return False

    # TODO: Update this once more than one extendee is allowed
    #  for spec in ext_spec.package.extendees:
    spec = ext_spec.package.extendee_spec
    if spec not in ext_spec:
        print "not actually an extension", ext_spec.name
        return False
```

Main issue is that package.is_extension checks whether the package *could be* an extension. In the case where I was trying to link in ```git```, the ```libxml2``` package conditionally extends python but in my case actually doesn't, so when you grab extendee_spec you're not actually getting an installed package.

scheibelp commented:
> In our case, the error occurred because the .extendee_spec of a given package was still present, even though the package was not configured to be an extension. If I understood the code correctly, this is not supposed to happen.

Oh I see what you're saying there. Yeah in this case that is more likely something that should be handled in package.py

scheibelp commented:
> When I build "git" in your PR and then try ```spack view soft sandbox/ git``` I get the error ```Specs passed to a DirectoryLayout must be concrete!.```

I edited my follow-up comment but I also wanted to make explicit that this is not an issue with this PR.

scheibelp commented:
> Upon (quick) second investigation it might also be the case that the flatten(...)-method should simply use concretized() instead…

It shouldn't need to call ```normalized``` or ```concretized``` (i.e. just use ```spec.traverse()```) since at that point the function is working with specs resolved with the ```disambiguate_specs``` function (which fetches installed specs from the database).

I can avoid the error I mentioned by updating Package. I'll try to get a PR out for it by the end of tomorrow.

scheibelp commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -454,10 +463,12 @@ def write_easy_install_pth(self, exts):
                         continue
 
-                    paths.append(line)
+                    line = os.path.realpath(line)
```
Some of the easy-install.pth files contain lines like:

```./Babel-2.3.4-py2.7.egg```

In which case realpath may expand relative to the current directory (vs. the directory where the easy-install.pth file is)

As you mention, it appears that https://github.com/LLNL/spack/pull/3587 will do away with this. I'll add a comment there that perhaps this entire function should disappear. If it can't for some reason then IMO this issue should be fixed.

scheibelp commented:
> I just noticed #3587, which solves the problem of filesystem views being unusable with python extensions in general,

I don't think #3587 fully resolves this, since you take the effort to use ```activate``` logic for extensions you benefit from whatever files they choose not to merge. In the case of python modules that includes "site.py"

I'm noticing an extra file that is added is ```.spack-empty``` - do you know where that comes from?

obreitwi commented:
Yeah, apparently there are still some cases left where `easy_install.pth`s need to be merged.

The `.spack-empty` token-files are so that `LinkTree` does not remove empty folders when unmerging (at least as far as I understood). 

On 14 April 2017 04:32:54 CEST, scheibelp <notifications@github.com> wrote:
>> I just noticed #3587, which solves the problem of filesystem views
>being unusable with python extensions in general,
>
>I don't think #3587 fully resolves this, since you take the effort to
>use ```activate``` logic for extensions you benefit from whatever files
>they choose not to merge. In the case of python modules that includes
>"site.py"
>
>I'm noticing an extra file that is added is ```.spack-empty``` - do you
>know where that comes from?
>
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/LLNL/spack/pull/3227#issuecomment-294071095

scheibelp commented:
When I merge https://github.com/LLNL/spack/pull/3853 into your PR I can add git to a view

> apparently there are still some cases left where `easy_install.pth`s need to be merged.

In that case, does my note about ```realpath``` make sense?

obreitwi commented:
Yes, we should just remove duplicate lines from the resulting
easy_install.pth, imho.

On Fri, Apr 14, 2017 at 05:33:04PM -0700, scheibelp wrote:
> When I merge https://github.com/LLNL/spack/pull/3853 into your PR I can add git to a view
> 
> > apparently there are still some cases left where `easy_install.pth`s need to be merged.
> 
> In that case, does my note about ```realpath``` make sense?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/LLNL/spack/pull/3227#issuecomment-294260740

scheibelp commented:
> Yes, we should just remove duplicate lines from the resulting easy_install.pth, imho.

To be clear I think there is an issue beyond duplicate lines. For example if I install py-babel and then run

```./bin/spack view soft sandbox/ py-babel```

With your PR, then ```sandbox/lib/python2.7/site-packages/easy-install.pth``` looks like:

```
/g/g17/scheibel/repos/spack/Babel-2.3.4-py2.7.egg
/g/g17/scheibel/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.7/py-pytz-2016.10-f54ik3bezdsnnyfvixrtvreb5t5vytbc/lib/python2.7/site-packages/pytz-2016.10-py2.7.egg
/g/g17/scheibel/repos/spack/pytz-2016.10-py2.7.egg
import sys; new=sys.path[sys.__plen:]; del sys.path[sys.__plen:]; p=getattr(sys,'__egginsert',0); sys.path[p:p]=new; sys.__egginsert = p+len(new)
```

The problematic line IMO is ```/g/g17/scheibel/repos/spack/Babel-2.3.4-py2.7.egg``` - that path doesn't exist. I'm pretty sure it looks like that because your PR calls ```realpath``` on a path which starts with ```./``` while Spack is being run from ```/g/g17/scheibel/repos/spack/```

obreitwi commented:
Yes, you are completely correct.

I removed the call to `realpath` and just make sure any duplicate lines are removed (plus a rebase on current `upstream/develop`).

scheibelp commented:
I'm aiming to review this more-fully be the end of 4/21: at the moment I'm tempted to request substantial refactoring (not issues that originate here but relate to prior technical debt that is wrapped up in your edits) but I'm trying to figure out how to avoid that.

obreitwi commented:
No problem, I initially tried to be as non-invasive with my changes as possible, but I am all for refactoring the code into one set of functions that handles both global extensions as well as extensions in filesystem views.

(Just removed my unnecessary use of `iteritems` in favor of `items` to maintain compatibility to `python3`.)

scheibelp commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -475,21 +485,46 @@ def write_easy_install_pth(self, exts):
+        # if we are activating in a filesystem view, we need to remove
+        # the link to the main easy_install.pth file prior to writing
+        # a custom one
+        if path_view:
```
I'm surprised this is required since the ```python_ignore``` function includes easy_install.pth in the list of files not to link

obreitwi replied:
But it only ignores `easy_install.pth` files under `bin/`. I remember I had the problem that the `easy_install.pth` file under `site-packages` was linked (which I think is correct behavior for globally activated extensions as it was before).

scheibelp replied:
> it only ignores easy_install.pth files under bin/

When you say that do you mean this pattern:

```r'bin/easy_install[^/]*$',```

Because a few lines above it I also see:

```r'site-packages/easy-install\.pth$'```

So as far as I can tell, this should already be taken care of (and for example I see the develop branch exclude easy_install.pth if I activate py-pytz). If you see that happen then I should probably investigate that.

> the easy_install.pth file under site-packages was linked (which I think is correct behavior for globally activated extensions as it was before).

As of now ```Python.write_easy_install_path``` in develop either deletes or overwrites the easy_install.pth file (since it opens in "w" mode). If it was accidentally linked in before then this would mitigate that, although it shouldn't have to.

obreitwi replied:
Mh… while I do fully agree with what you say I could have sworn I had the problems I described while writing the PR.

I guess I'll just remove change in question and once the [tests](https://github.com/LLNL/spack/pull/3227#discussion_r113430793) have been implemented we'll see if this is still an issue (which it should not be).

obreitwi commented:
I meant to get around to answer you today, but unfortunately I'll have to put it off till tomorrow!

obreitwi commented on lib/spack/spack/package.py:
```diff
@@ -1552,104 +1562,154 @@ def _sanity_check_extension(self):
-                            msg % (self.spec.short_spec, aspec.short_spec))
-
-        self.extendee_spec.package.deactivate(self, **self.extendee_args)
+                        if remove_dependents:
```
Because `force` (as far as I understand) simply means to not perform any checks whether the package in question is activated or if another acitvated package depends on it.

Up until now there was no support to automatically remove the dependents of a package as well. However, I do agree that the functionality could also be moved into `cmd/view.py`.

scheibelp replied:
Got it - I overlooked the fact that it was actually deactivating the dependents. I'm assuming this falls into the category of changes that you made that weren't essential to the core goal of this PR but annoyed you while you were working on it. The recursive deactivations initially made me worried but this looks OK. I'd like to see a test for it (although in general there's some discussion to be had about testing).

obreitwi replied:
Yeah, I meant to get around to write some tests if there was interest in this PR (which is, of course, everyone's excuse…) - unfortunately, I did not find any tests for the original `view` command implementation.

However, I think there definitely should be tests before this PR gets merged.

obreitwi commented on lib/spack/spack/package.py:
```diff
@@ -1552,104 +1562,154 @@ def _sanity_check_extension(self):
                                   (self.name, self.extendee.name))
 
-    def do_activate(self, force=False):
+    def do_activate(self, path_view="", force=False, verbose=True):
```
I agree, I changed it back to `None` but kept the `if path_view` checks.

obreitwi commented on lib/spack/spack/cmd/view.py:
```diff
@@ -179,13 +189,58 @@ def flatten(seeds, descend=True):
+        linked_spec = spec_in_view(spec, path)
+        if linked_spec == spec:
+            pass
+        else:
```
Agreed!

obreitwi commented on lib/spack/spack/package.py:
```diff
@@ -1872,6 +1932,14 @@ def _hms(seconds):
     return ' '.join(parts)
 
 
+def _format_pw(path_view, post=""):
```
Agreed.

obreitwi commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -432,11 +437,15 @@ def python_ignore(self, ext_pkg, args):
 
-    def write_easy_install_pth(self, exts):
+    def write_easy_install_pth(self, exts, prefix=None):
+        if prefix is None:
```
Whoops, missed that…

obreitwi commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -475,21 +485,46 @@ def write_easy_install_pth(self, exts):
+            exts_global = spack.store.layout.extension_map(self.spec,
+                                                           path_view="")
+            # since extension_map() returns a copy, we can update directly
+            exts.update(exts_global)
```
I fully agree, I wanted to keep the possibility of having some extensions activated globally for all file system views, however, I have never used that feature and do agree that it would be much cleaner to make it either/or.

In that case: What is the correct behavior in the case of creating a filesystem views while globally activated extensions are already present? Because the extensions' files have already been linked to the python-package's install directory.
* Do you simply check for globally activated packages and fail if there are any?
* Do you filter the python-package's directory for links?

scheibelp replied:
> What is the correct behavior in the case of creating a filesystem views while globally activated extensions are already present? Because the extensions' files have already been linked to the python-package's install directory.

I prefer your first approach (fail if anything is globally activated). I think people are going to be attracted to one method over the other. If people get excited about using both then a different PR could add a phase that temporarily deactivates all extensions.

obreitwi replied:
Okay, I'll get started to move the per-view extensions into their own `ViewDirectoryLayout` inheriting from `DirectoryLayout` then.

obreitwi commented on lib/spack/spack/directory_layout.py:
```diff
@@ -88,34 +89,38 @@ def check_installed(self, spec):
 
-    def extension_map(self, spec):
-        """Get a dict of currently installed extension packages for a spec.
+    def extension_map(self, spec, path_view=""):
```
I am all for it: In fact, that was the first approach I considered, but since it prevented globally activated extensions to propagate into filesystem views I decided against it.

However, if we decide against globally activated extensions (which I would prefer!) this would be a non-issue and I'd get started right away.

scheibelp replied:
>However, if we decide against globally activated extensions (which I would prefer!) this would be a non-issue and I'd get started right away.

To be clear, when you say that do you mean "not mixing global and view extensions" rather than "disabling support for global extensions altogether"? If so then I agree, and I think that is consistent with the conversation at https://github.com/LLNL/spack/pull/3227/files#r113287441

scheibelp commented:
I wanted to make sure: as far as I can tell you didn't have any additional questions WRT my requests. Let me know if I missed something - thanks!

A couple tests that may be worthwhile:

* Check if write_easy_install_pth creates the file correctly (maybe if this were refactored into a helper method that output a list of paths it would be easier to test)
* Check that you cant activate two different versions of the same extension into a view

FWIW there is a mock packages repo in ```var/spack/repos/builtin.mock/packages/``` where you can add made up packages for testing. ```test/install.py``` contains examples which do the installations (and also have examples of creating temporary directories which you could use to test views out).

That's pretty terse so I'm happy to discuss it more, but just some initial thoughts on it.

obreitwi commented:
Sorry for my delayed response, I updated the PR to my curent progress, which is
as follows (spoilers: no tests yet, unfortunately).

When activating/deactivating extensions, we still need information from the
original layout (namely source directories from where to link, i.e.
install-prefixes). Therefore, I implemented a `ViewYamlDirectoryLayout` that is
meant to replace the default YamlDirectoryLayout in-place at
`spack.store.layout`, keeping code duplication to a minimum. However, that also
means that most of the logic implementing filesystem views still resides in
`cmd/view.py` (which is appropriate for the time being, if you ask me).

I also had to introduce a new `layout.get_extendee_target_directory(…)` method
so that extendees can figure out where they should link extension-files to.

One problem I have not been able to resolve yet is an unconcretized
`.extendee_spec` when activating an extension. It arose after rebasing onto the
current `upstream/develop` branch, but I have been unable to find its cause
yet. If I understood you correctly, `spec.traverse()` should take care of the
concretization, correct? Because it is the `spec` found during DAG-traversal
that is not concrete. This appears to occur whether or not I call
`.concretized()` on user supplied specs in the `flatten(…)`-function in
`cmd/view.py`.

Another point that still bothers me is that there currently is no method of
handling duplicated files. Granted, most of the time is is the exact same file
that two packages install on their own, so a simple `--ignore-conflicts` will
suffice. However, there should be a way of resolving/merging common duplicate
file patterns. One very simple example are `./usr/share/info/dir`-files that
need to be merged via `install-info`.

And finally, yes, the tests are still missing and we should not merge this PR
before they are done, however, I would appreciate your feedback regarding the
proposed solution.

scheibelp commented on lib/spack/spack/cmd/view.py:
```diff
@@ -290,14 +451,48 @@ def visitor_statlink(specs, args):
+    # now that we have loaded all necessary specs from the database we can
+    # monkey patch the directory layout so that all activation mechanisms will
+    # use the view based approach instead
+    setattr(spack.store, "layout", layout)
```
I'm imagining to make this cleaner (IMO), the DirectoryViewLayout could be replaced by a ViewExtensionLayout, the point being that all current logic is located in the same place, but the object you created is responsible entirely for handling extensions (in a sense it implements an interface that only handles the extension logic). The existing YamlDirectoryLayout could also implement these functions.

* Extract ```DirectoryLayout._extension_map``` into a standalone function (I think accepting ```all_specs``` as a parameter to construct ```specs_by_hash``` would work for this).
* ```Package.activate``` now takes an ExtensionLayout (either the ViewExtensionLayout or the YamlDirectoryLayout) so spack.store.layout is always referring to the spack installation prefix for the package, and the ExtensionLayout provides the target directory to symlink into
* ViewExtensionLayout and YamlDirectoryLayout both offer functions to: retrieve an extension map for a spec; provide the target prefix to activate an extension spec in
* The ```spack activate``` command would then provide the YamlDirectoryLayout implementation of the implicit ExtensionLayout interface to ```Package.activate```

In this approach, you don't need to replace ```spack.store.layout``` and the ViewExtensionLayout wouldn't extend YamlDirectoryLayout so there wouldn't be confusion over what functions are needed from it.

Although this is a significant shift in ideas, my intent is that it is not a large change in terms of adding or removing code - not that this is necessarily the thing to optimize, but I'm saying this so if you get a different impression then I should probably explain myself better.

If this is unclear or disagreeable let me know (I'm assuming this explanation is insufficient).

scheibelp commented:
> One problem I have not been able to resolve yet is an unconcretized .extendee_spec when activating an extension. It arose after rebasing onto the current upstream/develop branch, but I have been unable to find its cause yet.

I'll look into this tomorrow

> If I understood you correctly, spec.traverse() should take care of the concretization, correct? Because it is the spec found during DAG-traversal that is not concrete. This appears to occur whether or not I call .concretized() on user supplied specs in the flatten(…)-function in cmd/view.py

```traverse``` by itself does not concretize, but as you say you are calling concretized so the extendee_spec should be set. 

Could you let me know where this is occurring? In that case, is the extendee activated conditionally, i.e. ```extends(..., when=)```?

> However, there should be a way of resolving/merging common duplicate file patterns. One very simple example are ```./usr/share/info/dir-files``` that need to be merged via install-info.

If I understand correctly, are these expected to overlap (i.e. some documentation files are duplicated in different python modules)? Perhaps an interim solution is to add them to ```ignore```. Perhaps there should be a new filter which places the files if they don't exist but doesn't care if they do. Or maybe a check could be added which compares the target and source files and doesn't complain as long as they match. I have to think on this more. If you have more examples that would be great.

scheibelp commented:
> One problem I have not been able to resolve yet is an unconcretized .extendee_spec when activating an extension. It arose after rebasing onto the current upstream/develop branch, but I have been unable to find its cause yet.

Would you be able to let me know which package this occurs for? At least I tried it for ```py-pytz``` and did not get this error.

One thing I did notice when trying ```spack view soft sandbox/ py-pytz``` was that the ```pytz``` module files were showing up in ```sandbox/.spack/python/lib/python2.7/site-packages/pytz``` (where I would instead expect them to show up in ```sandbox/lib/python2.7/site-packages/pytz```)

obreitwi commented:
I think I might have fallen into the "never reuse old packages when
updating spack"-trap. After clearing all installed packages the bug
seems to have vanished, but I am currently trying to finish the
refactoring prior to investigating further.

On Wed, May 10, 2017 at 07:58:04PM -0700, scheibelp wrote:
> > One problem I have not been able to resolve yet is an unconcretized .extendee_spec when activating an extension. It arose after rebasing onto the current upstream/develop branch, but I have been unable to find its cause yet.
> 
> Would you be able to let me know which package this occurs for? At least I tried it for ```py-pytz``` and did not get this error.
> 
> One thing I did notice when trying ```spack view soft sandbox/ py-pytz``` was that the ```pytz``` module files were showing up in ```sandbox/.spack/python/lib/python2.7/site-packages/pytz``` (where I would instead expect them to show up in ```sandbox/lib/python2.7/site-packages/pytz```)
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/LLNL/spack/pull/3227#issuecomment-300669092

obreitwi commented:
Okay, finally managed to finish the refactoring.

As you suggested, I split out the extension functionality into a separate `ExtensionLayout` residing at `spack.store.extensions`. It still gets automatically replaced whenever `view`-operations are performed, but in an automated fashion (I really did not want to pass the `ExtensionLayout` around everywhere).

Furthermore, I finally got around to bundling all filesystem view in a single `FilesystemView`/`YamlFilesystemView`-object, which allowed me to reduce the size of `cmd/view.py` tremendously.

… tests are still missing, though …

scheibelp commented:
Thanks for all the work on this! I'll look through it by Monday 5/15

obreitwi commented:
Okay, thanks, in the meantime I'll see to it that Travis has nothing to complain about.

scheibelp commented:
Generally I like the reorganization but I'm still seeing the issue I mentioned earlier:

> One thing I did notice when trying ```spack view soft sandbox/ py-pytz``` was that the ```pytz``` module files were showing up in ```sandbox/.spack/python/lib/python2.7/site-packages/pytz``` (where I would instead expect them to show up in ```sandbox/lib/python2.7/site-packages/pytz```)

obreitwi commented:
Oh, I guess I couldn't see the forest for the trees. All extensions were linked into the meta-folder of the extendee instead of the root of the view. Should be fixed now.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -325,6 +326,12 @@ def ancestor(dir, n=1):
     return parent
 
 
+def assuredir(path):
```
Can this be removed? As far as I can tell, this function is no longer called based on the updates in this PR.

obreitwi replied:
Oh, yes, I moved it from `cmd/view.py` to utils and then forgot to delete it.

scheibelp commented on lib/spack/llnl/util/link_tree.py:
```diff
@@ -112,4 +126,6 @@ def unmerge(self, dest_root, **kwargs):
                     raise ValueError("%s is not a link tree!" % dest)
-                os.remove(dest)
+                # only remove if src actually points to dest
+                if filecmp.cmp(src, dest, shallow=True):
```
Would it be clearer to use ```if os.path.realpath(dest) == src```? IMO it is sufficient if ```dest``` is in fact a symlink to ```src```; if you agree with me on that IMO this is more straightforward since it takes more effort to understand the implications of ```filecmp.cmp```.

obreitwi replied:
Using `realpath` would be sufficient if we only ever used symlinks, but for hardlinks the `realpath`s differ.

Although I do admit that this should only ever be an issue when adding (and subsequently removing) packages containing conflicting files with hardlinks. In all other cases the original `os.remove` would have been sufficient.

scheibelp replied:
> I updated the comment accordingly.

Sounds good - thanks!

scheibelp commented on lib/spack/llnl/util/link_tree.py:
```diff
@@ -62,10 +63,16 @@ def find_conflict(self, dest_root, **kwargs):
+           Note that files blocking directories will still cause an error.
+        """
         kwargs['order'] = 'pre'
+        report_existing = kwargs.get("report_existing", False)
```
IMO this should also be called ```ignore_conflicts```. The variable definitely results in reporting existing files, but IMO the most important effect is that this option allows for conflicts to occur.

obreitwi replied:
I don't know why I did not choose that name, but I renamed it now…

obreitwi commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -0,0 +1,494 @@
+        to_keep = all_specs - to_deactivate
+
+        # remove all packages depending on the ones to remove
+        to_deactivate.update(find_dependents(to_keep, to_deactivate))
```
Up until now, there was no check on packages depending on the ones to remove. There was a simple `--dependencies` switch (enabled by default), that would remove dependencies of packages to remove.

However, if you removed the dependency of another package, rendering it unusable, there was no check performed and hence no warning.

scheibelp replied:
Got it. I would prefer there would be an option for this like ```--remove-dependents```. If it is true by default, warn the user when you are automatically removing dependents; if it is false by default, warn the user when you are leaving "detached" dependents around.

obreitwi replied:
Okay, I added the option (default True), so that the user does not have to worry about removing the dangling packages and the view remains usable. People who claim to know what they are doing should put in the extra work :)

obreitwi commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -0,0 +1,494 @@
+
+        return True
+
+    def activate_standalone(self, spec):
```
When I decided on the name I wanted to keep in the spirit of extensions in that activating a package is linking it somewhere else, but I have no strong opinion on the issue.

Your concern is very valid that it might confuse people → I'll change it.

obreitwi commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -0,0 +1,494 @@
+        raise NotImplementedError
+
+
+def with_extensions(f):
```
Hm… I wanted to mirror the "global" behavior of the `DirectoryLayout`/databases etc. Why should extensions be passed explicitly but everything else remains in global objects residing under `spack.store` - seems a bit asymmetric.

My actual primary motivation was to keep `Package.activated` a property, but you are right, explicit is better than implicit.

scheibelp commented:
> I'll also start thinking of tests to suggest and I'll add an update on that by the end of tomorrow.

* Firstly I suggested testing ```write_easy_install_pth``` at https://github.com/LLNL/spack/pull/3227#issuecomment-298143184. This might be easier to test if that function were refactored to take the names of the .pth files to merge, and returned the contents of the merged file (or if that were to be placed in a helper method).
* Some existing mock packages already create installation files. You could test installing and then adding any of them to a view.
* There are a couple of mock extension packages, but neither of them define multiple versions. IMO you could modify either one to have 2 versions, install both, and then confirm that both cannot be added to a view.
* Install a dependent, add it (plus the dependency) to a view, remove the dependency from the view. The functionality to decide on what gets removed could be refactored into its own function to make this trivial (so you don't necessarily have to install things, just make sure the specs are calculated properly).

I'll keep thinking on it. Let me know if you think these aren't useful or aren't in scope for your PR. Also if you think the tests are useful but want to chat about how to implement them let me know.

obreitwi commented:
Okay, just updated the branch. Now global state is not modified, the `status` and `remove` command are a bit more lenient when it comes to specifying specs (basically names are enough) and the color-output is disabled when there is no TTY attached to stdout.

After making my head hurt at first (I wanted to not remove dependencies that have non-user-specified dependents) I decided to make it nice and simple: 
* If dependencies are to be removed they are added to the list of specs to remove.
* If dependents are to be removed all dependents are identified and removed as well (this includes dependents of dependencies of user-specified packages). A warning is printed.
* If dependents are not to be removed, we identify them anyway and print a warning to the user (even without verbose output) that the packages will become unusable.

Basically I don't see users removing packages from views that much anyway. It is just too easy to delete everything and add exactly what you want. Now it is the most conservative approach (which happens to delete more rather than having unusable packages linger).

If the PR is fine otherwise I'll have a closer look at the tests in the coming days.

scheibelp commented on lib/spack/spack/database.py:
```diff
@@ -720,14 +720,17 @@ def installed_dependents(self, spec):
         Return the specs of all packages that extend
         the given spec
         """
+        # TODO: Shouldn't this function be called "activated_extensions_for"?
```
Yes - if you could change the name I'd appreciate it.

obreitwi replied:
Done

scheibelp commented on lib/spack/llnl/util/link_tree.py:
```diff
@@ -112,4 +126,7 @@ def unmerge(self, dest_root, **kwargs):
                 if not os.path.islink(dest):
                     raise ValueError("%s is not a link tree!" % dest)
-                os.remove(dest)
+                # only remove if src actually points to dest
```
IIRC dest is actually the file/symlink in the view, so IMO it would be clearer if this read "if dest points to src" or perhaps even better is "only remove if dest is a hardlink/symlink to src" and it occurs to me that it would also be useful to explain how this can occur since it is not expected to be a common situation. So in all: 

"remove if dest is a hardlink/symlink to src; this will only be false if two packages are merged into a prefix and have a conflicting file"

obreitwi replied:
Done

scheibelp commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -0,0 +1,516 @@
+                tty.warn(self._croot +
+                         "The following dependents will be removed: %s"
+                         % ", ".join((s.name for s in dependents)))
+                to_deactivate.update(dependents)
```
Since you go through the work of aggregating dependents here, my understanding is that if you sorted them based on a post-order traversal of the dependency DAG that calling ```Package.do_deactivate``` with ```with_dependents=True``` would not be required. I don't think there's any harm in doing it this way although the redundancy may be confusing. If I'm right I think it would make sense to make a comment on it (I don't necessarily think it requires a change).

obreitwi replied:
Yeah, you are correct: Prior to this PR there was an exception thrown if an activated dependent was detected, so I added the option to remove dependents instead (back when this PR was a quick hack).

I am unsure as to whether the method you describe will be reliable enough (I have yet to fully understand all code related to the DAG), but I'll add a comment that this should be done in the future if performance is too detrimental.

scheibelp commented on lib/spack/spack/directory_layout.py:
```diff
@@ -149,6 +118,67 @@ def remove_install_directory(self, spec):
+        raise NotImplementedError
+
+    @property
+    def hidden_file_paths(self):
```
I don't think either implementation of the interface implements this method. Is there some reason it is here?

obreitwi replied:
Oh, right, that was erroneously copied from `DirectoryLayout`, removed.

scheibelp commented:
> Basically I don't see users removing packages from views that much anyway. It is just too easy to delete everything and add exactly what you want.

Makes sense to me

> If the PR is fine otherwise I'll have a closer look at the tests in the coming days.

Added a few more comments. IMO they are all minor issues.

obreitwi commented:
After having had a quick look at the testing routines, some questions/thoughts came up:

* I think `write_easy_install_pth` should be tested and refactored out; especially given the recent rewrites (#3569, #3583). Although, partly thanks to #3587, it is not that much of an issue and really outside the scope of this PR (which is being able to activate different sets of extensions from the same spack-repository into different filesystem views). It was its primary motivation, though.

* The `builtin.mock`-python package is missing its `do_activate` function. How should proper functionality of the actual package be verified in the current testing framework?

* On what level should be tested? Only user based input with `builtin.mock` as repository or directly on the level of `FilesystemView`? In the latter case, I am torn between adding some packages to `builtin.mock` and having mocked `Spec`s and `Package` generated on the fly (as with the install command tests) that just install an adjustable set of random-dummy files based on their version-hash.

* Even more general: Do we have to test for proper linking of files given the fact that `LinkTree` is already covered by tests? Would testing conflict-detection, dependency-resolutoin and proper removal of dependencies/dependents be enough when performed on "specs without install files"-level?

* If linking should be tested, are their testing routines that allow `builtin.mock` to contain installed specs (so that these could actually be linked to a temporary filesystem view)?

scheibelp commented:
> The builtin.mock-python package is missing its do_activate function. How should proper functionality of the actual package be verified in the current testing framework?

IMO you don't need to test ```Python.activate```. Even testing ```write_easy_install_pth``` is a bit beyond the norm. That being said I think you can access the "real" python easily if you don't use the ```install_mockery``` fixture.

> Do we have to test for proper linking of files given the fact that LinkTree is already covered by tests?

IMO no

> Would testing conflict-detection, dependency-resolution and proper removal of dependencies/dependents be enough when performed on "specs without install files"-level? 

I could see how the last two could be tested on specs without install files. The first one would be harder assuming we're talking about file conflict detection (vs. detecting whether some other version of a package has been added to a view), although IMO you don't have to work with specs (could just be one of the link_tree tests)

> On what level should be tested? Only user based input with builtin.mock as repository or directly on the level of FilesystemView? 

If by "Only user based input with builtin.mock" you mean invocations of cmd.view I don't think it has to be at that level. Direct testing of filesystem_view seems good to me. 

> In the latter case, I am torn between adding some packages to builtin.mock and having mocked Specs and Package generated on the fly (as with the install command tests) that just install an adjustable set of random-dummy files based on their version-hash.

While ```test/cmd/install``` uses mock Specs and packages ```test/install``` does not. IMO builtin.mock would be easier here given that the view logic calls more functions but either way would be fine.

mathstuf commented:
I've rebased this and split it into multiple commits in #5415.

alalazo commented:
#5415 went in. Can we close this?

tgamblin commented:
Superseded by #5415
